src/Umbraco.Cms.Api.Management/Services/Signs/HasPendingChangesSignProvider.cs

```diff
@@ -0,0 +1,49 @@
+﻿using Umbraco.Cms.Api.Management.Mapping.Content;
+using Umbraco.Cms.Api.Management.ViewModels;
+using Umbraco.Cms.Api.Management.ViewModels.Document;
+using Umbraco.Cms.Api.Management.ViewModels.Document.Collection;
+using Umbraco.Cms.Api.Management.ViewModels.Tree;
+using Umbraco.Cms.Core;
+using Umbraco.Cms.Core.Models;
+using Umbraco.Cms.Core.Services;
+
+namespace Umbraco.Cms.Api.Management.Services.Signs;
+
+public class HasPendingChangesSignProvider : ISignProvider
+{
+    private const string Alias = Constants.Conventions.Signs.Prefix + "PendingChanges";
+
+    private readonly IContentService _contentService;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="HasPendingChangesSignProvider"/> class.
+    /// </summary>
+    public HasPendingChangesSignProvider(IContentService contentService) => _contentService = contentService;
+
+    public bool CanProvideSigns<TItem>()
+        where TItem : IHasSigns =>
+        typeof(TItem) == typeof(DocumentTreeItemResponseModel) ||
+        typeof(TItem) == typeof(DocumentCollectionResponseModel);
+
+    public Task PopulateSignsAsync<TItem>(IEnumerable<TItem> itemViewModels)
+        where TItem : IHasSigns
+    {
+        foreach (TItem item in itemViewModels)
+        {
+            IContent? content = _contentService.GetById(item.Id);
```

Comment:
We need to try to find a more efficient way to do this, as it suffers currently from an "N+1" problem.  I.e. for every item in the collection we are making a request to get the content, which will be going to the database.

If you look at the `HasScheduleSignProvider`, we are getting all the information we need for all entity IDs, in one database hit.  And that's what we need to try to do here.

I see `IContentService` has a `GetByIds` method, so likely we should use that.

Looking at `DocumentVariantStateHelper` it does appear we need to retrieve the full details of content items, and so there's likely not much more optimization we can do in terms of bringing back a lighter model.